lib/Dialect/FIRRTL/Transforms/AssignOutputDirs.cpp

```diff
@@ -0,0 +1,158 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/AnnotationDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLInstanceGraph.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+
+#define DEBUG_TYPE "firrtl-assign-output-dirs"
+
+using namespace circt;
+using namespace firrtl;
+namespace path = llvm::sys::path;
+namespace fs = llvm::sys::fs;
+
+using hw::OutputFileAttr;
+
+// If moduleOutputDir is a relative path, convert it to an absolute path, by
+// interpreting moduleOutputDir as relative to the outputDir.
+static void makeAbsolute(StringRef outputDir,
+                         SmallString<64> &moduleOutputDir) {
+  fs::make_absolute(outputDir, moduleOutputDir);
+  path::remove_dots(moduleOutputDir, true);
+}
+
+// If outputDir is a prefix of moduleOutputDir, then make moduleOutputDir
+// relative to outputDir. Otherwise, leave moduleOutputDir as absolute.
+static void tryMakeRelative(StringRef outputDir,
+                            SmallString<64> &moduleOutputDir) {
+  if (moduleOutputDir.starts_with(outputDir))
+    moduleOutputDir.erase(moduleOutputDir.begin(),
+                          moduleOutputDir.begin() + outputDir.size());
+}
+
+static void makeCommonPrefix(SmallString<64> &a, StringRef b) {
+  // truncate 'a' to the common prefix of 'a' and 'b'.
+  size_t i = 0;
+  size_t e = std::min(a.size(), b.size());
+  for (; i < e; ++i)
+    if (a[i] != b[i])
+      break;
+  a.resize(i);
+
+  // truncate 'a' so it ends on a directory seperator.
+  auto sep = path::get_separator();
+  while (!a.empty() && !a.ends_with(sep))
+    a.pop_back();
+}
+
+static void makeCommonPrefix(StringRef outputDir, SmallString<64> &a,
+                             OutputFileAttr attr) {
+  if (attr) {
+    SmallString<64> b(attr.getFilename());
+    makeAbsolute(outputDir, b);
+    makeCommonPrefix(a, b);
+  } else {
+    makeCommonPrefix(a, outputDir);
+  }
+}
+
+static OutputFileAttr getOutputFile(Operation *op) {
+  return op->getAttrOfType<hw::OutputFileAttr>("output_file");
+}
+
+namespace {
+struct AssignOutputDirsPass
+    : public AssignOutputDirsBase<AssignOutputDirsPass> {
+  AssignOutputDirsPass() = default;
+  AssignOutputDirsPass(StringRef outputDir) : AssignOutputDirsPass() {
+    outputDirOption = std::string(outputDir);
+  }
+
+  void runOnOperation() override;
+};
+} // namespace
+
+void AssignOutputDirsPass::runOnOperation() {
+  SmallString<64> outputDir(outputDirOption);
+  if (fs::make_absolute(outputDir)) {
+    llvm::errs()
```

Comment:
Passes should prefer to emit diagnostics rather than printing to `llvm::errs()` themselves.

This is a bit unusual as its a pass option so doesn't obviously have a location to report on-- maybe report on the current operation or if that's strange, maybe `UnknownLoc`? WDYT?